compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -2101,7 +2101,9 @@ trait Applications extends Compatibility {
             else defn.FunctionOf(commonParamTypes, WildcardType)
           overload.println(i"pretype arg $arg with expected type $commonFormal")
           if (commonParamTypes.forall(isFullyDefined(_, ForceDegree.flipBottom)))
-            withMode(Mode.ImplicitsEnabled)(pt.typedArg(arg, commonFormal))
+            withMode(Mode.ImplicitsEnabled) {
```

Comment:
At first glance this looks too simplistic. We do a complicated dance not to overcommit by caching arguments that can be passed to several different expected types. How can we be sure that we can simply cache in this case? Maybe we can, but we'd need a detailed comment why.

Reply:
Yes, a comment is deserved here. Currently I see the rationale is this: the `expected type` for pretyping here come from all the overloading candidates and they are all fully defined.

> I just checked: it does not work with master anymore. I think #12138 is to blame.

I rebased against master, it seems to work.